Add ISocketConnectionContextFactory

HaoK opened the pull request:
Fixes https://github.com/dotnet/aspnetcore/issues/34522

Still trying to figure out why shutdown tests are failing, but want to see a full ci run

HaoK committed:
Initial surgery

HaoK committed:
Initial surgery

HaoK committed:
Update SocketConnectionListener.cs

HaoK committed:
Try to fix shutdown tests

HaoK committed:
Fix bind tests

HaoK committed:
Update SocketConnectionListener.cs

kevin-montrose commented on src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionContextFactory.cs:
```diff
@@ -0,0 +1,68 @@
+    /// <summary>
+    /// A factory for socket based connections contexts.
+    /// </summary>
+    internal sealed class SocketConnectionContextFactory : ISocketConnectionContextFactory, IAsyncDisposable
```
Should be `public`, yes?

davidfowl replied:
Why?

kevin-montrose replied:
I might be misunderstanding, but I thought that's what [the public API entry here](https://github.com/dotnet/aspnetcore/pull/34769/files#diff-8b15fbe07f965adfec63a9c4ca239ff43dd1310dcd9d6f8bd69b49a78ca76fb9R9) indicated.

I also don't quite follow how'd you get a handle on this, to use separately from `SocketTransportFactory`.

davidfowl replied:
Dependency injection

kevin-montrose replied:
So, call `UseSockets()`, grab this factory, and then remove both the registered `SocketTransportFactory` and this factory?  Feels a little weird to register both, when presumably a custom transport is almost always gonna be registered.

Also seems kind of brittle if `UseSockets()` registers anything beyond those two in the future.

davidfowl commented on src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs:
```diff
@@ -171,12 +174,6 @@ public ValueTask DisposeAsync()
             _memoryPool.Dispose();
 
-            // Dispose any pooled senders
-            foreach (var setting in _settings)
```
Where did this go?

HaoK replied:
I moved this into the dispose of the new factory: https://github.com/dotnet/aspnetcore/blob/7465741d00784841555dbb5a59671d8b414340dc/src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionContextFactory.cs#L63

The pooled sender/settings and transport schedule bits definitely feel a bit wrong still

davidfowl replied:
I think we need to add a singleton one to the container but we can't use it in the transport. We need to create one per listener in the transport.

davidfowl commented:
This changes the memory pool and socketsender lifetime. We still want one of those per endpoint.

HaoK commented:
Guessing that's why the shutdown tests are all failing :)  Yeah the analogy I used during standup today is that I feel like I've  transplanted a kidney into a lung cavity, still figuring out how stuff works, thanks for the early feedback.

HaoK commented:
> This changes the memory pool and socketsender lifetime. We still want one of those per endpoint.

It wasn't clear to me how best to flow what is in the settings to the new SocketConnectionContextFactory when the API only had `Create(Socket, SocketConnectionOptions)`. internal properties on the SocketConnectionOptions?

HaoK committed:
Fix lifetime issues with internal properties

HaoK committed:
Update WebHostBuilderSocketExtensions.cs

HaoK committed:
Make context factory public and fix UseKestrel

HaoK committed:
Make context factory public and fix UseKestrel

HaoK committed:
Fix DelaySocketOperations to match NoDelay defaults

HaoK commented:
Looks like all the tests are passing now except for the unix domain sockets, not sure what's different about those

HaoK committed:
Don't access socket.NoDelay cuz unix sockets

HaoK committed:
Remove debugging

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionContextFactory.cs:
```diff
@@ -0,0 +1,52 @@
+        {
+            _transportOptions = transportOptions.Value;
+
+            _transportScheduler = _transportOptions.UnsafePreferInlineScheduling ? PipeScheduler.Inline :
```
Since this is important, we should probably just add UnsafePreferInlineScheduling to SocketConnectionOptions. It would be weird to change per socket, but I don't think we should prevent it. Just make sure we still reuse the same IOQueue if it switches back and forth.

HaoK replied:
> Just make sure we still reuse the same IOQueue if it switches back and forth.

I'm not sure I follow what additional guard we need here?  I moved the scheduler to the ConnectionOptions for now, the factory is nice and simple now, it basically just passes everything from the options to SocketConnection

halter73 replied:
The problem with just moving the scheduler is that now people using the SocketConnectionContextFactory can no longer use the IOQueue since it's internal and we very much want them to by default for performance. I guess we could make it public and nullable and use the IOQueue when we normally would if it's null.

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionContextFactory.cs:
```diff
@@ -0,0 +1,52 @@
+                options.SenderPool,
+                options.InputOptions,
+                options.OutputOptions,
+                waitForData: _transportOptions.WaitForDataBeforeAllocatingBuffer);
```
```suggestion
                options.WaitForDataBeforeAllocatingBuffer);
```

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs:
```diff
@@ -117,28 +119,33 @@ internal void Bind()
 
                     // Only apply no delay to Tcp based endpoints
+                    var noDelay = false;
+// REVIEW: should this move to inside of _contextFactory.Create??
```
I think since SocketConnectionContextFactory.Create() takes a Socket, there's no reason to. They could set Socket.NoDelay themselves before calling Create().

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs:
```diff
@@ -117,28 +119,33 @@ internal void Bind()
-                        waitForData: _options.WaitForDataBeforeAllocatingBuffer);
+                    var connectionOptions = new SocketConnectionOptions()
+                    {
+                        DelaySocketOperations = !noDelay,
```
`_options.NoDelay` is different from `DelaySocketOperations` which was referring to this https://github.com/dotnet/aspnetcore/pull/34639/files#diff-6c2098732ca5c5b26b7451144b6f78147bcbad8d1b444a02b568ba575e9317f3R39. This *is* confusing though. We should probably come up with a new name for the new uption. @davidfowl

HaoK replied:
Okay, good idea to rename :)

halter73 commented on src/Servers/Kestrel/Kestrel/src/WebHostBuilderKestrelExtensions.cs:
```diff
@@ -33,6 +33,7 @@ public static IWebHostBuilder UseKestrel(this IWebHostBuilder hostBuilder)
             {
                 // Don't override an already-configured transport
                 services.TryAddSingleton<IConnectionListenerFactory, SocketTransportFactory>();
+                services.TryAddSingleton<ISocketConnectionContextFactory, SocketConnectionContextFactory>();
```
I don't think we need to register this as a service in UseKestrel(). This is for other custom transports (IConnectionListenerFactory implementations) to use, not us.

halter73 replied:
Though this does ensure we have decent test coverage. I'm just wary of adding new extensibility points. Maybe it's fine.

halter73 replied:
The more I think about this, the more I think the extra test coverage is not a reason to do this. Most of the logic is in SocketConnection which is shared anyway. We can add tests specifically for any interesting behavior in SocketConnectionContextFactory if need be.

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs:
```diff
@@ -24,17 +23,20 @@ internal sealed class SocketConnectionListener : IConnectionListener
         internal SocketConnectionListener(
             EndPoint endpoint,
             SocketTransportOptions options,
+            ISocketConnectionContextFactory contextFactory,
```
The intention behind this API was not to allow people to create their own ConnectionContexts from Sockets, it was to give other Socket-based custom transports an easier time creating ConnectionContexts. We should continue to use SocketConnection directly in SocketConnectionListener.

HaoK commented on src/Servers/Kestrel/Transport.Sockets/src/IISocketConnectionContextFactory.cs:
```diff
@@ -0,0 +1,22 @@
+    /// <summary>
+    /// Defines an interface that provides the mechanisms to create a socket based <see cref="ConnectionContext"/>.
+    /// </summary>
+    public interface ISocketConnectionContextFactory
```
One less interface sounds good to me, this doesn't seem like it will be a very commonly used extensibility point right?  Nuke the interface then?

HaoK committed:
CR feeedback

HaoK committed:
Cleanup

HaoK commented:
I'm assuming I should also update https://github.com/dotnet/aspnetcore/blob/bdafb7a6c7445414a543467d6c7850fa968c3b6d/src/Servers/Kestrel/Transport.Sockets/src/Client/SocketConnectionFactory.cs#L74 to use this new abstraction?  At this point I might well update the SocketConncection ctor to just take the new options class

HaoK committed:
Fix test

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionOptions.cs:
```diff
@@ -0,0 +1,49 @@
+
+        internal PipeScheduler Scheduler { get; init; } = default!;
+
+        internal ISocketsTrace Trace { get; init; } = default!;
```
It seems wrong to mark these as non-nullable. Any non-internal user of this API will leave these null, right?

halter73 replied:
I don't like having internal properties at all. External callers `ISocketConnectionContextFactory` won't be able to use these. The reason for making this a service instead of just a public static method is so these pools could be maintained internally by the SocketConnectionContextFactory.

HaoK replied:
Yeah these internal properties feel weird, they are only here for us (and they should never be null when used), totally agree that this feels a bit yuck.  Any ideas on an alternative to shuttle this info?

halter73 replied:
Our built-in stuff should just continue new up SocketConnection directly so we can pass in these internal types. The SocketConnectionContextFactory can create an `ISocketsTrace` from an `ILoggerFactory` that's constructor injected like SocketTransportFactory does.

halter73 replied:
For the pools, the SocketConnectionContextFactory can just keep these in private fields. The reason it's a service and not a static method is not so it can be injested by the socket transport, but so it can keep these pools around between calls.

halter73 commented:
> I'm assuming I should also update [SocketConnectionFactory.cs] to use this new abstraction?

I don't think so. This abstraction is for use by third-party transports not for allowing people to create their own ConnectionContexts from Sockets. I don't think anyone asked for the latter, so I'd rather not support that if we don't have to. I don't think SocketTransportFactory should consume the (I)SocketConnectionContextFactory either.

HaoK commented:
>  I don't think anyone asked for the latter, so I'd rather not support that if we don't have to. I don't think SocketTransportFactory should consume the (I)SocketConnectionContextFactory either.

Okay I defer to you guys, just from the general purpose naming its a bit surprising that we wouldn't use it everywhere we create SocketConnectionContexts

halter73 commented:
> Okay I defer to you guys, just from the general purpose naming its a bit surprising that we wouldn't use it everywhere we create SocketConnectionContexts

That's partly why I like removing the interface and not registering it with the IServiceCollection in UseKestrel(). In many cases, I expect it won't be registered as a service at all, but instead newed-up by a custom IConnectionListenerFactory with services it gets in its constructor.

When I filed the API proposal, I was considering an interface and not the concrete class. So we'll need to add the constructor to the API proposal. I think the ILoggerFactory would be the only parameter though.

HaoK commented:
> I don't think SocketTransportFactory should consume the (I)SocketConnectionContextFactory either

Right now the transport factory is who new's up the SocketConnectionListener in Bind so if it doesn't use it, no-one will since the socket connection listener isn't currently getting created from DI no?

halter73 commented:
Correct. This isn't for us to use. It's for third-party transports to use. So we'll definitely want to add some tests to make sure it all works end-to-end.

HaoK commented:
> That's partly why I like removing the interface and not registering it with the IServiceCollection in UseKestrel(). In many cases, I expect it won't be registered as a service at all, but instead newed-up by a custom IConnectionListenerFactory with services it gets in its constructor.

What's so bad about just treating this as a normal service that is responsible for creating ConnectionContexts from a Socket and SocketConnectionOptions? I'm obviously coming at this from a totally naïve viewpoint...that's what the current naming would have me expect its responsibility is

HaoK committed:
Remove interface

HaoK committed:
Remove interface

halter73 commented:
We have a lot more flexibility in the future we we don't consume the SocketConnectionContextFactory as a service in the socket transport.

If we consume it as a service, that effectively forces us to change a public API any time we want to change the SocketConnection constructor. Sometimes this might be warranted anyway so third-party transports using it always have the same capabilities we do, but that might not always be necessary if we just continue calling the constructor directly with the internal pools we manage and whatnot. I like being able to make this decision on a case-by-case basis rather than be force into it.

Anyone who's willing to do all the work to create a completely custom ConnectionContext from a Socket should just register their own IConnectionListenerFactory and bypass the socket transport completely since they've already done 90% of the work of implementing a custom IConnectionListenerFactory anyway. If someone just wants to decorate the ConnectionContext the socket transport would normally give you, they can use connection middleware and it'd work with any transport.

halter73 commented:
I've updated the issue with my current thinking about how the API should look. I should have never made it an interface or called it a service. The reason we decided not to make it a static method as originally proposed is not so it can be replaced and injested by the socket transport, but so it can keep pools around between calls.

HaoK committed:
Updates

HaoK committed:
Update to new API

HaoK committed:
Undo

HaoK commented on src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionOptions.cs:
```diff
@@ -0,0 +1,39 @@
+        /// <remarks>
+        /// Defaults to false.
+        /// </remarks>
+        public bool DeferFirstOperation { get; set; }
```
Note: this isn't currently used at all

halter73 replied:
Let's just remove it. I think we're going to revert #34458 and close #34639 in favor of this PR.

HaoK replied:
Cool, removed

HaoK committed:
Revert

HaoK commented on src/Servers/Kestrel/Transport.Sockets/src/PublicAPI.Shipped.txt:
```diff
@@ -18,6 +18,6 @@ Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketTransportOptions.Uns
 Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketTransportOptions.WaitForDataBeforeAllocatingBuffer.set -> void
 ~Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketTransportFactory.BindAsync(System.Net.EndPoint endpoint, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) -> System.Threading.Tasks.ValueTask<Microsoft.AspNetCore.Connections.IConnectionListener>
-~Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketTransportFactory.SocketTransportFactory(Microsoft.Extensions.Options.IOptions<Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketTransportOptions> options, Microsoft.Extensions.Logging.ILoggerFactory loggerFactory) -> void
+~Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketTransportFactory.SocketTransportFactory(Microsoft.Extensions.Options.IOptions<Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketTransportOptions!>! options, Microsoft.Extensions.Logging.ILoggerFactory! loggerFactory) -> void
```
My build fails without this update, but I vaguely remember Doug saying something about not changing public/shipped apis

HaoK replied:
Ok found it, guess this is supposed to be in the public unshipped

HaoK committed:
Undo submodule update

HaoK committed:
Undo api changes

HaoK committed:
Update PublicAPI.Shipped.txt

HaoK committed:
Remove DeferFirstOperation

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs:
```diff
@@ -42,45 +38,37 @@ internal sealed class SocketConnectionListener : IConnectionListener
-                _settingsCount = ioQueueCount;
-                _settings = new Settings[_settingsCount];
+                _factoryCount = ioQueueCount;
+                _factories = new SocketConnectionContextFactory[_factoryCount];
```
I think it would be better to have the `SocketConnectionContextFactory` manage its on set of `IOQueues` rather than forcing any implementer of a custom Socket-based `IConnectionListenerFactory` to basically do this themselves or suffer worse perf.

I know @davidfowl said he didn't think we should have `SocketConnectionOptions.IOQueueCount` though which we would need if `SocketConnectionContextFactory` manages its on set of `IOQueues`. Maybe we should set up a meeting with the three of us.

HaoK committed:
Start connection in Create

HaoK committed:
Move sender pool queues into Factory

HaoK committed:
Cleanup

HaoK commented:
Pretty clean now, basically net changes look like we moved a bunch of code from the listener -> factory and split the transport options into two classes. Socket tests are green locally so I think this is ready for final review @halter73 @davidfowl

HaoK committed:
Rename field

davidfowl commented on src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs:
```diff
@@ -12,24 +10,8 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets
-        /// <remarks>
-        /// Defaults to <see cref="Environment.ProcessorCount" /> rounded down and clamped between 1 and 16.
-        /// </remarks>
-        public int IOQueueCount { get; set; } = Math.Min(Environment.ProcessorCount, 16);
```
I wonder if this is a breaking change?

HaoK replied:
Code check said no as well: Detected modification to baseline API files. PublicAPI.Shipped.txt files should only be updated after a major release. See /docs/APIBaselines.md for more information.

So back this out and just duplicate this in a new options class?

davidfowl replied:
Yep

halter73 replied:
I don't think introducing a new base class by itself is a breaking change, but I'm pretty sure moving properties down to the new base class is. https://docs.microsoft.com/en-us/dotnet/core/compatibility/ says "Introducing a new base class" "REQUIRES JUDGMENT".

HaoK replied:
Well I reverted that and just added an internal copy constructor to flow things to the 'base' ContextFactory options

HaoK replied:
So we could do this in a non breaking way by making the properties virtual and override them to call base or something?

davidfowl replied:
What you have is fine. Leave them separate.

HaoK committed:
CR feedback

HaoK committed:
Fix test

davidfowl reviewed (approved):
This is super clean.

davidfowl commented:
OK @kevin-montrose we re-did the change and make a class that can be created (without DI) that has all of the implementation guts of the transport. To enable your scenario, you can write a custom transport that handles the listening logic on both the incoming server and the server being routed to.

@mconnew for your scenario, we exposed the pipe settings so you can control the thresholds. We don't let you change the IDuplexPipe implementation before we read from the socket though. Hopefully that isn't a huge blocker. This should enable you to write a custom transport in CoreWCF without losing the behavior or performance enhancements we continue to make in the transport.
